Initial implementation of margin-trim for legacy inline layout in non-paginated and single column block containers

sammygill opened the pull request:
#### 6e0475da2ecd191b009ae9f1d1d728e7dfcd8972
<pre>
Initial implementation of margin-trim for legacy inline layout in non-paginated and single column block containers
<a href="https://bugs.webkit.org/show_bug.cgi?id=252062">https://bugs.webkit.org/show_bug.cgi?id=252062</a>
rdar://problem/105283580

Reviewed by NOBODY (OOPS!).

This patch attempts to cover as many scenarios as possible for margin
trim in regards to legacy line layout and serves as a basis for an
initial implementation for the property in this context. There should be a comprehensive
suite of tests that range from simple cases to most complex, but there
will likely need to be some iteration on this patch. For example,
currently imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all.html
does not render quite correctly and the underlying cause needs to
be investigated. Paginated and multi-column content support will also
need to be provided in a further patch.

When floats are being placed, we can check to see if we can trim some
of its margins based on its surrounding constraints. In particular, we
can check to see if there are any other items on the left/right
(depending on the position of the float) of the box at that candidate
vertical position. If there is nothing there (i.e. the float would be
placed flush against its containing block), then we can trim the
respective margin. However, if there is an item at that vertical, then
we must wait until the float gets its final position before tirmming
any of its margins. This is because trimming a margin too early could
result it in being placed in a position where it should not be.

&lt;div style=&quot;margin-trim: inline;&quot;&gt;
  &lt;div style=&quot;float: left; width: 50px; height: 50px; background-color: green; margin-inline-start: 30px;&quot;&gt;&lt;/div&gt;
  &lt;div style=&quot;float: right; width: 50px; height: 50px; background-color: red; margin-inline-end: 30px;&quot;&gt;&lt;/div&gt;
&lt;/div&gt;

In this simple example we can easily trim both the margins of the floats
as we attempt to place them. Inside BreakingContext::handleFloat() we
check the logical offset for the line against the logical offset for
content inside the containing block. If these two values are equal,
then the float would be placed against the box and we can trim the
margin early.

In all cases this may not entirely be possible:
&lt;style&gt;
container {
    display: block;
    width: 110px;
    margin-trim: inline;
    font-size: 0px;
}
item {
    display: block;
    width: 100px;
    height: 25px;
    background-color: green;
}
.inline-block {
    display: inline-block;
    height: 50px;
}
.float-left {
    float: left;
}
.overflowing {
    margin-inline-start: 55px;
}
&lt;/style&gt;
&lt;container&gt;
    &lt;item class=&quot;inline-block&quot;&gt;&lt;/item&gt;
    &lt;item class=&quot;float-left&quot;&gt;&lt;/item&gt;
    &lt;item class=&quot;float-left overflowing&quot;&gt;&lt;/item&gt;
&lt;/container&gt;
For the last float we see that there is content on the left side at
this vertical position, so we do not trim the margin here. Instead we
have to wait until this float gets pushed under all the other ones
before trimming.

If we trim too early, the float could get positioned in a spot where it
would not normally fit with its margin. For this we fall back to the
code in RenderBlockFlow::computeLogicalLocationForFloat that will trim
the margin at its final position. This is the same logic that was used
to trim the margins for floats in block containers that did not have
any inline children.

* LayoutTests/TestExpectations:
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-up-to-content-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-up-to-content-inline-layout-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-up-to-content-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-start-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-start-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-nested-within-span-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-nested-within-span-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-nested-within-span.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-with-bidi-content-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-with-bidi-content-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-with-bidi-content.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vertl-lr-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vertl-lr.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-nested-within-span-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-nested-within-span-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-nested-within-span.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-bidi-content-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-bidi-content-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-bidi-content.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-intruding-floats-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-intruding-floats-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-intruding-floats.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-word-break-all-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-word-break-all-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-word-break-all.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-min-content-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-min-content-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-multiple-formatting-contexts-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-multiple-formatting-contexts-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-nested-inside-inline-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-nested-inside-inline-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-inline-layout-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-nested-within-span-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-nested-within-span-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-nested-within-span.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-with-bidi-content-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-with-bidi-content-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-with-bidi-content.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vertl-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vertl-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-nested-within-span-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-nested-within-span-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-nested-within-span.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-bidi-content-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-bidi-content-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-bidi-content.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-intruding-floats-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-intruding-floats-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-intruding-floats.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-multiple-formatting-contexts-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-multiple-formatting-contexts-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-nested-inside-inline-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-nested-inside-inline-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout.html: Added.
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::canUseForChild):
* Source/WebCore/rendering/LegacyLineLayout.cpp:
(WebCore::LegacyLineLayout::contentBoxLogicalHeight const):
* Source/WebCore/rendering/LegacyLineLayout.h:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock):
(WebCore::RenderBlockFlow::computeInlinePreferredLogicalWidths const):
(WebCore::RenderBlockFlow::blockFormattingContextInFlowContentHeight const):
(WebCore::RenderBlockFlow::blockFormattingContextInFlowBlockLevelContentHeight const): Deleted.
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::BreakingContext::handleFloat):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e0475da2ecd191b009ae9f1d1d728e7dfcd8972

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108791 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41624 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/116 "Updated wpe dependencies (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118093 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112673 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9167 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101018 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114559 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14514 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97733 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42484 "Found 2 new test failures: webgl/2.0.y/conformance/glsl/bugs/character-set.html, webgl/2.0.y/conformance/glsl/misc/fragcolor-fragdata-invariant.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96502 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29371 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84334 "Found 2 new API test failures: /TestWebKit:WebKit.OnDeviceChangeCrash, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10674 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30722 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11435 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7650 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16822 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50313 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13022 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/line/BreakingContext.h:
```diff
@@ -364,12 +367,36 @@ inline void BreakingContext::handleOutOfFlowPositioned(Vector<RenderBox*>& posit
+        return m_width.fitsOnLineExcludingTrailingWhitespace(widthWitoutMargin.toFloat());
+    };
+    auto hasIntrudingFloatOnSide = [&](UsedFloat floatPosition) {
+        return (floatPosition == UsedFloat::Left) ? m_block.logicalLeftOffsetForLine(m_block.logicalHeight(), DoNotIndentText)  != m_block.logicalLeftOffsetForContent(m_block.logicalHeight()) :  m_block.logicalRightOffsetForLine(m_block.logicalHeight(), DoNotIndentText) != m_block.logicalRightOffsetForContent(m_block.logicalHeight());
```
I'd rename this function as the term "intrusive floats" are reserved for cases when they overlap the block container they live in and intrude into sibling block containers (and by that they potentially constrain lines in those sibling block containers)

alanbaradlay replied:
what I think you essentially check here is whether there's already a float on a particular logical y position (or range).

sammygill replied:
Changed

alanbaradlay commented on LayoutTests/fast/inline/legacy-margin-trim/block-container-float-multiple-formatting-contexts-inline-layout.html:
```diff
@@ -0,0 +1,36 @@
+    float: left;
+}
+span {
+    display: inline-block;
```
this is rather misleading. what's the point of making a <span> and turn it into an inline-block?

sammygill replied:
Yeah not sure why I did that. Should be changed now

alanbaradlay commented on LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout.html:
```diff
@@ -0,0 +1,33 @@
+<container>
+    <item style="float: left; background-color: green; margin-inline-start: 20px;"></item>
+    <item class="tall" style="float: right; background-color: red;"></item>
+    <item style="float: left; background-color: blue; margin-inline-start: 110px;"></item>
```
this does not look like inline content.

sammygill replied:
Added some text

alanbaradlay commented on LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-multiple-formatting-contexts-inline-layout.html:
```diff
@@ -0,0 +1,35 @@
+span {
+    display: inline-block;
+    width: 25px;
+}
```
Same as above, not sure why <span> is turned into inline-block.

sammygill replied:
Should be changed now

alanbaradlay commented on LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout.html:
```diff
@@ -0,0 +1,32 @@
+    <item style="float: left; background-color: green; margin-inline-start: 20px;"></item>
+    <item class="tall" style="float: right; background-color: red;"></item>
+    <item style="float: left; background-color: blue; margin-inline-start: 110px;"></item>
+</container>
```
3 floats in a block box. Not sure what we test here in the context of this patch.

sammygill replied:
Removed the first float and added some text. Blue float should be placed right under the text

sammygill committed:
Initial implementation of margin-trim for legacy inline layout in non-paginated and single column block containers
https://bugs.webkit.org/show_bug.cgi?id=252062
rdar://problem/105283580

Reviewed by NOBODY (OOPS!).

This patch attempts to cover as many scenarios as possible for margin
trim in regards to legacy line layout and serves as a basis for an
initial implementation for the property in this context. There should be a comprehensive
suite of tests that range from simple cases to most complex, but there
will likely need to be some iteration on this patch. For example,
currently imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all.html
does not render quite correctly and the underlying cause needs to
be investigated. Paginated and multi-column content support will also
need to be provided in a further patch.

When floats are being placed, we can check to see if we can trim some
of its margins based on its surrounding constraints. In particular, we
can check to see if there are any other items on the left/right
(depending on the position of the float) of the box at that candidate
vertical position. If there is nothing there (i.e. the float would be
placed flush against its containing block), then we can trim the
respective margin. However, if there is an item at that vertical, then
we must wait until the float gets its final position before tirmming
any of its margins. This is because trimming a margin too early could
result it in being placed in a position where it should not be.

<div style="margin-trim: inline;">
  <div style="float: left; width: 50px; height: 50px; background-color: green; margin-inline-start: 30px;"></div>
  <div style="float: right; width: 50px; height: 50px; background-color: red; margin-inline-end: 30px;"></div>
</div>

In this simple example we can easily trim both the margins of the floats
as we attempt to place them. Inside BreakingContext::handleFloat() we
check the logical offset for the line against the logical offset for
content inside the containing block. If these two values are equal,
then the float would be placed against the box and we can trim the
margin early.

In all cases this may not entirely be possible:
<style>
container {
    display: block;
    width: 110px;
    margin-trim: inline;
    font-size: 0px;
}
item {
    display: block;
    width: 100px;
    height: 25px;
    background-color: green;
}
.inline-block {
    display: inline-block;
    height: 50px;
}
.float-left {
    float: left;
}
.overflowing {
    margin-inline-start: 55px;
}
</style>
<container>
    <item class="inline-block"></item>
    <item class="float-left"></item>
    <item class="float-left overflowing"></item>
</container>
For the last float we see that there is content on the left side at
this vertical position, so we do not trim the margin here. Instead we
have to wait until this float gets pushed under all the other ones
before trimming.

If we trim too early, the float could get positioned in a spot where it
would not normally fit with its margin. For this we fall back to the
code in RenderBlockFlow::computeLogicalLocationForFloat that will trim
the margin at its final position. This is the same logic that was used
to trim the margins for floats in block containers that did not have
any inline children.

* LayoutTests/TestExpectations:
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-up-to-content-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-up-to-content-inline-layout-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-end-up-to-content-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-start-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-block-start-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout-vert-lr.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-nested-within-span-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-nested-within-span-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-nested-within-span.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-with-bidi-content-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-with-bidi-content-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-end-with-bidi-content.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vertl-lr-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout-vertl-lr.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-nested-within-span-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-nested-within-span-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-nested-within-span.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-bidi-content-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-bidi-content-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-bidi-content.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-intruding-floats-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-intruding-floats-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-start-with-intruding-floats.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-word-break-all-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-word-break-all-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-inline-word-break-all.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-min-content-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-min-content-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-multiple-formatting-contexts-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-multiple-formatting-contexts-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-nested-inside-inline-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-nested-inside-inline-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-ref.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout-expected.html: Added.
* LayoutTests/fast/inline/legacy-margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-inline-layout-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-end-up-to-content-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-block-start-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-nested-within-span-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-nested-within-span-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-nested-within-span.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-with-bidi-content-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-with-bidi-content-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-end-with-bidi-content.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vert-lr-rtl.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vertl-lr-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout-vertl-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-nested-within-span-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-nested-within-span-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-nested-within-span.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-bidi-content-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-bidi-content-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-bidi-content.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-intruding-floats-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-intruding-floats-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-start-with-intruding-floats.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-inline-word-break-all.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-min-content-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-multiple-formatting-contexts-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-multiple-formatting-contexts-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-nested-inside-inline-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-nested-inside-inline-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-margin-has-margin-trimmed-at-final-position.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-overflowing-positioned-at-first-available-vertical-position-with-trimmed-margin-inline-layout.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-float-trimmed-margin-allows-float-to-fit-inline-layout.html: Added.
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::canUseForChild):
* Source/WebCore/rendering/LegacyLineLayout.cpp:
(WebCore::LegacyLineLayout::contentBoxLogicalHeight const):
* Source/WebCore/rendering/LegacyLineLayout.h:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock):
(WebCore::RenderBlockFlow::computeInlinePreferredLogicalWidths const):
(WebCore::RenderBlockFlow::blockFormattingContextInFlowContentHeight const):
(WebCore::RenderBlockFlow::blockFormattingContextInFlowBlockLevelContentHeight const): Deleted.
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::BreakingContext::handleFloat):
